lib/phoenix_live_view/channel.ex

```diff
@@ -1554,8 +1554,22 @@ defmodule Phoenix.LiveView.Channel do
   end
 
   defp authorize_session(%Session{} = session, endpoint, %{"url" => url}) do
+    %Session{view: view, live_session_name: session_name} = session
+
     if Session.main?(session) do
-      {:ok, session, session_route(session, endpoint, url), url}
+      # Ensure the session's LV module and live session name still match on connect.
+      # If the route has changed the LV module or has moved live sessions, the client
+      # will fallback to full page redirect to the current URL.
+      case session_route(session, endpoint, url) do
+        %Route{view: ^view, live_session: %{name: ^session_name}} = route ->
+          {:ok, session, route, url}
+
+        %Route{} ->
+          {:error, :stale}
+
+        nil ->
+          {:error, :stale}
```

Comment:
Right above we return :unauthorized, maybe return the same here for consistency?

```suggestion
          {:error, :unauthorized}

        nil ->
          {:error, :unauthorized}
```

Also, do we care about logging something? We also do above and I think they would be triggered in similar scenarios.